tests/test_warm_reboot.py

```diff
@@ -1651,8 +1682,11 @@ def test_system_warmreboot_neighbor_syncup(dvs, testlog):
     # should finish the store within 10 seconds
     time.sleep(10)
 
-    check_kernel_reachable_v4_neigh_num(dvs, NUM_OF_NEIGHS)
-    check_kernel_reachable_v6_neigh_num(dvs, NUM_OF_NEIGHS)
+    check_kernel_reachable_v4_neigh_num(dvs, NUM_OF_NEIGHS/2)
+    check_kernel_reachable_v6_neigh_num(dvs, NUM_OF_NEIGHS/2)
+
+    check_kernel_stale_v4_neigh_num(dvs, NUM_OF_NEIGHS/2)
+    check_kernel_stale_v6_neigh_num(dvs, NUM_OF_NEIGHS/2)
```

Comment:
With half of the neighbor entries in stale state, it seems del operations be triggered against appDB/asicDB?  
  ```
# check syslog and sairedis.rec file for activities
    check_syslog_for_neighbor_entry(dvs, marker, 0, 0, "ipv4")
     check_syslog_for_neighbor_entry(dvs, marker, 0, 0, "ipv6")
     (nadd, ndel) = dvs.CountSubscribedObjects(pubsub)
     assert nadd == 0
     assert ndel == 0
```

Reply:
No, "stale" means the entry was not removed from kernel.  only when the entry got expired/deleted, then we will see the netlink messages and trigger appDB/asicDB changes.

Reply:
With https://github.com/Azure/sonic-swss/blob/master/neighsyncd/neighsyncd.cpp#L61 , only valid entries will be taken into account.  I guess the reconciliation logic will get rid of the stale entries?  That should be the expected behavior.

Reply:
I see.  
To confirm my understanding of the behavior:  The original state for arp/nd neighbor injected to kernel is stale, there is arp/ns packet generated immediately for each entry, upon receiving reply from neighbor the state will move to reachable.  For entries without reply,  they stay in stale state until gc_stale_time times out.  It seems 60 seconds is large enough, but it was changed to 180 seconds in your test cases.  Did you see time taken longer than that for the operations of kernel entry injection and arp/ns packet sending?  I also saw you reduced the NUM_OF_NEIGHS to 128 from 512.

Reply:
A few things here:
1, 60 seconds stale timer is fine for vs docker on my machine.  However, for slow machines we don't know how long it takes for the process between neigh insert and the neigh verification logic in the test code.  The time took for neigh insertion and arp/ns packets was small (a few seconds) even for 1K entries, the time spent was more for the the neighbor verification logic, the sleep, and ping etc...in the testing code.  And to make the vs testing stable, I increased the stale timer. 

2, In terms of NUM_OF_NEIGHS,  to support higher number of neighbors at vs docker and have expected behavior, we need tune the kernel parameters "gc_thresh1/2/3" at host side of the vs docker.  
The thresholds will determine when the gc task to run,  the gc task could probe and delete the stale entries by kernel.  If gc didn't run and stale timer could kick in and expire the stale entries.
Also, if the number of neighs is bigger than the hardware limit (gc_thresh3), system is going to a crazy state (vs test will hang).
It worked fine on my setup for 512 and 1K entries after tuning them.  but to avoid doing the tuning on different machines with even different kernels(I am not sure how to tune them in the test code itself), I simply reduced the number for vs test cases

Anyways,  The critical time of the restore process is small.  The timing issues and behavior issue we are dealing here are pretty much relevant to vs docker test cases.  And if we need more entries to run on SONiC (more than 1K per family), we will have to tune those kernel variables as well on the box.